toolchain/sem_ir/value_stores.h

```diff
@@ -146,17 +146,35 @@ class ConstantValueStore : public Yaml::Printable<ConstantValueStore> {
   llvm::SmallVector<ConstantId, 0> values_;
 };
 
-// Provides storage for instructions representing global constants.
+// Provides storage for instructions representing deduplicated global constants.
 class ConstantStore {
  public:
-  // Add a constant instruction.
-  auto Add(InstId inst_id) -> void { values_.push_back(inst_id); }
+  explicit ConstantStore(SemIR::File* sem_ir, llvm::BumpPtrAllocator& allocator)
+      : allocator_(&allocator), constants_(sem_ir) {}
 
-  auto array_ref() const -> llvm::ArrayRef<InstId> { return values_; }
-  auto size() const -> int { return values_.size(); }
+  // Adds a new constant instruction, or gets the existing constant with this
+  // value. Returns the instruction ID and a bool indicating whether a new
+  // instruction was added.
+  auto GetOrAdd(Inst inst) -> std::pair<SemIR::InstId, bool>;
+
+  auto vector() const -> std::vector<InstId>;
```

Comment:
Largely because this isn't supposed to be used anywhere performance- or size-sensitive, but I suppose we're generally using `SmallVector<T, 0>` for that these days. Switched.

Reply:
I would even say `SmallVector<T>` is appropriate if you're saying performance doesn't matter.